library/std/src/env.rs

```diff
@@ -318,22 +318,21 @@ impl Error for VarError {
 ///
 /// # Safety
 ///
-/// Even though this function is currently not marked as `unsafe`, it needs to
-/// be because invoking it can cause undefined behaviour. The function will be
-/// marked `unsafe` in a future version of Rust. This is tracked in
-/// [rust#27970](https://github.com/rust-lang/rust/issues/27970).
-///
 /// This function is safe to call in a single-threaded program.
 ///
-/// In multi-threaded programs, you must ensure that are no other threads
-/// concurrently writing or *reading*(!) from the environment through functions
-/// other than the ones in this module. You are responsible for figuring out
-/// how to achieve this, but we strongly suggest not using `set_var` or
-/// `remove_var` in multi-threaded programs at all.
+/// This function is also always safe to call on Windows, in single-threaded
+/// and multi-threaded programs.
 ///
-/// Most C libraries, including libc itself do not advertise which functions
-/// read from the environment. Even functions from the Rust standard library do
-/// that, e.g. for DNS lookups from [`std::net::ToSocketAddrs`].
+/// In multi-threaded programs on other operating systems, you must ensure that
+/// environment through functions or global variables other than the ones in
+/// this module. These operating systems unfortunately provide no thread-safe
+/// way to mutate the environment, so you are responsible for figuring out how
+/// to achieve this, carefully controlling all the code that runs in all
+/// concurrent threads. We strongly suggest not using `set_var` or `remove_var`
+/// in multi-threaded programs at all. Most C libraries, including libc itself,
+/// do not advertise which functions read from the environment. Even functions
+/// from the Rust standard library do that, e.g. for DNS lookups from
+/// [`std::net::ToSocketAddrs`].
```

Comment:
Something went wrong here with the suggestion it seems, this isn't even a sentence any more.
```suggestion
/// In multi-threaded programs on other operating systems, we strongly suggest
/// not using `set_var` or `remove_var` at all.
/// The exact requirement is: you must ensure that
/// there are no other threads concurrently writing or *reading*(!) from the
/// environment through functions or global variables other than the ones in
/// this module.
/// The problem is that these operating systems do not provide a thread-safe
/// way to read the environment, and most C libraries, including libc itself,
/// do not advertise which functions read from the environment. Even functions
/// from the Rust standard library may read the environment without going
/// through this module, e.g. for DNS lookups from
/// [`std::net::ToSocketAddrs`]. No stable guarantee is made about which functions
/// may read from the environment in future versions of a library.
/// All this makes it not practically possible for you to guarantee that no other thread
/// will read the environment, so the only safe option is to not use `set_var` or
/// `remove_var` in multi-threaded programs at all.
```

Reply:
We currently guarantee it and I don't think we should remove that. This change is already disruptive enough as it is, let's avoid making it any worse than absolutely necessary.

Reply:
Do we gain something from keeping this lock in the documentation and thus likely forever in the code? I'd say that realistically, almost all Rust code calling this cannot guarantee that all environment access goes through these functions, but might think it can. As such, it might be negative to document it this way.

Changed it to your suggestion anyway.

Reply:
> Do we gain something from keeping this lock in the documentation and thus likely forever in the code?

It should probably be in the code permanently as a basic security mitigation. I think it's unreasonable to consider programs that need super-fast environment access; environment access should generally be hoisted as far as possible to the beginning of a program. So I don't really see that there is a downside here.

But for applications that do need do so some environment modification, we are either making it harder for things to go wrong, or automating adding the lock that a pure-Rust program would need.